src/NHibernate/Dialect/Function/BitwiseFunctionOperation.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections;
+using NHibernate.Engine;
+using NHibernate.SqlCommand;
+using NHibernate.Type;
+
+namespace NHibernate.Dialect.Function
+{
+	/// <summary>
+	/// Treats bitwise operations as SQL function calls.
+	/// </summary>
+	[Serializable]
+	public class BitwiseFunctionOperation : ISQLFunction
+	{
+		private readonly string _functionName;
+
+		/// <summary>
+		/// Creates an instance of this class using the provided function name
+		/// </summary>
+		/// <param name="functionName">
+		/// The bitwise function name as defined by the SQL-Dialect
+		/// </param>
+		public BitwiseFunctionOperation(string functionName)
+		{
+			_functionName = functionName;
+		}
+
+		#region ISQLFunction Members
+
+		public IType ReturnType(IType columnType, IMapping mapping)
+		{
+			return NHibernateUtil.Int64;
+		}
+
+		public bool HasArguments
+		{
+			get { return true; }
+		}
+
+		public bool HasParenthesesIfNoArguments
+		{
+			get { return true; }
+		}
+
+		public SqlString Render(IList args, ISessionFactoryImplementor factory)
+		{
+			var filteredArgs = new Queue();
```

Comment:
Why would you need queue here?

Reply:
I have done the fix in a "minimal implementation changes" way, mainly fixing the fact that putting some state without any locks in an instance shared by many threads cannot be thread safe. So I have kept the previous logic using a queue.